examples/author_disambiguation.py

```diff
@@ -0,0 +1,162 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of Beard.
+# Copyright (C) 2014 CERN.
+#
+# Beard is a free software; you can redistribute it and/or modify it
+# under the terms of the Revised BSD License; see LICENSE file for
+# more details.
+
+"""Simplified author disambiguation example.
+
+This example shows how to use semi-supervised block clustering for the author
+disambiguation problem. To goal is to cluster together all (author name,
+affiliation) tuples that correspond to the same actual person.
+
+.. codeauthor:: Gilles Louppe <g.louppe@cern.ch>
+
+"""
+
+import numpy as np
+import re
+
+from functools import wraps
+from sklearn.cross_validation import train_test_split
+
+from beard.clustering import BlockClustering
+from beard.clustering import ScipyHierarchicalClustering
+from beard.metrics import paired_f_score
+from beard.utils.strings import asciify
+
+
+def memoize(func):
+    """Memoization function."""
+    cache = {}
+
+    @wraps(func)
+    def wrap(*args):
+        if args not in cache:
+            cache[args] = func(*args)
+        return cache[args]
+
+    return wrap
+
+
+RE_NORMALIZE_LAST_NAME = re.compile("\s+|\-")
+RE_NORMALIZE = re.compile("(,\s(i|ii|iii|iv|v|vi|jr))|[\.'\-,]|\s+")
+
+
+@memoize
+def normalize(name):
+    """Transliterate a name to ascii and remove all special characters."""
+    name = asciify(name).lower()
+
+    try:
+        names = name.split(",", 1)
+        name = "%s, %s" % (RE_NORMALIZE_LAST_NAME.sub("", names[0]), names[1])
+    except:
+        pass
+
+    name = RE_NORMALIZE.sub(" ", name)
+    name = name.strip()
+
+    return name
+
+
+@memoize
+def initials(name):
+    """Compute the set of initials of a given name."""
+    return set([w[0] for w in name.split()])
+
+
+def affinity(X):
```

Comment:
Yes, possibly. Right now this function is really a heuristic which is specific to the data of this very example. This is why I leave it here. In the midterm however, we will have helpers to directly learn such functions.